Android/Samples/libraries/kml/src/main/java/com/microsoft/maps/kml/KMLParser.java

```diff
@@ -82,30 +88,92 @@ MapElementLayer internalParse(@NonNull String kml)
       mParser.setInput(stream, null);
       mParser.nextTag();
       mNameSpace = mParser.getNamespace();
-      parseOuterLayers();
+      parseOuterLayer();
+      applyStyles();
     }
     return mLayer;
   }
 
-  private void parseOuterLayers() throws IOException, XmlPullParserException, KMLParseException {
+  private void parseOuterLayer() throws IOException, XmlPullParserException, KMLParseException {
     while (moveToNext() != XmlPullParser.END_TAG) {
       if (mParser.getEventType() != XmlPullParser.START_TAG) {
         continue;
       }
       String type = mParser.getName();
-      if (type.equals("Placemark")) {
-        parseNameAndShape();
-      } else if (type.equals("Document") || type.equals("Folder")) {
-        parseOuterLayers();
+      switch (type) {
+        case "Placemark":
+          parseNameAndShape();
+          break;
+        case "Style":
+          parseStyleAddToMapStylesHolders();
+          break;
+        case "Document":
+        case "Folder":
+          parseOuterLayer();
+          break;
+        default:
+          skipToEndOfTag();
+          break;
+      }
+    }
+  }
+
+  private void parseStyleAddToMapStylesHolders()
+      throws XmlPullParserException, IOException, KMLParseException {
+    String id = mParser.getAttributeValue(null, "id");
+    if (mMapStylesHolders.containsKey(id)) {
+      throw new KMLParseException(
+          "Error at: " + mParser.getPositionDescription() + " ID " + id + " already seen.");
+    }
+    StylesHolder stylesHolder = new StylesHolder();
+    while (moveToNext() != XmlPullParser.END_TAG) {
+      if (mParser.getEventType() != XmlPullParser.START_TAG) {
+        continue;
+      }
+      String type = mParser.getName();
+      if (type.equals("IconStyle")) {
+        parseIconStyle(stylesHolder);
+      } else {
+        skipToEndOfTag();
+      }
+    }
+    mMapStylesHolders.put(id, stylesHolder);
+  }
+
+  private void parseIconStyle(@NonNull StylesHolder stylesHolder)
+      throws XmlPullParserException, IOException, KMLParseException {
+    while (moveToNext() != XmlPullParser.END_TAG) {
+      if (mParser.getEventType() != XmlPullParser.START_TAG) {
+        continue;
+      }
+      String type = mParser.getName();
+      if (type.equals("Icon")) {
+        parseIcon(stylesHolder);
       } else {
         skipToEndOfTag();
       }
     }
   }
 
+  private void parseIcon(@NonNull StylesHolder stylesHolder)
+      throws XmlPullParserException, IOException, KMLParseException {
+    while (moveToNext() != XmlPullParser.END_TAG) {
+      if (mParser.getEventType() != XmlPullParser.START_TAG) {
+        continue;
+      }
+      String type = mParser.getName();
+      if (type.equals("href")) {
+        String url = parseText();
+        InputStream inputStream = new URL(url).openConnection().getInputStream();
```

Comment:
This does a synchronous network operation which can be slow. Please document on parse that it should not be called on the UI thread if the KML file may contain any references to external resources and that they will be downloaded synchronously.